src/compiler/AST/Common/Type.elm

```diff
@@ -55,7 +63,71 @@ toString type_ =
             "Bool"
 
         List param ->
-            "List " ++ toString param
+            "List " ++ wrapParens param dict
 
         Unit ->
             "()"
+
+
+getTypeVariablesIndex : Type -> Dict Int String
+getTypeVariablesIndex type_ =
+    getTypeVariablesIndexHelp type_ (typeVariablesLetters, Dict.empty)
+    |> Tuple.second
+
+
+getTypeVariablesIndexHelp : Type -> (List String, Dict Int String) -> (List String, Dict Int String)
+getTypeVariablesIndexHelp type_ (freeLetters, dict) =
+    case type_ of
+        Var int ->
+            case freeLetters of
+                [] ->
+                    ([], dict)
+
+                hd :: xs ->
+                    if Dict.member int dict then
+                        (freeLetters, dict)
+                    else
+                        (xs, Dict.insert int hd dict)
+
+        Function t1 t2 ->
+            getTypeVariablesIndexHelp t1 (freeLetters, dict)
+            |> getTypeVariablesIndexHelp t2
```

Comment:
this is totally a nitpick but I think writing it as
```elm
(freeLetters, dict)
  |> getTypeVariablesIndexHelp t1
  |> getTypeVariablesIndexHelp t2
```
would better show the symmetry - that it recurses for each of its child expressions.